### ISSUE TITLE
gridstack.all 1.x import  failure

### ISSUE THREAD

**Issue by fuyouping** [gridstack/gridstack.js#1243]
vue-element-admin  import  gridstack.all  failure

gridstack.js 1.1.1

in page.vue

import GridStack from '../../plug-in/gridstack/dist/gridstack.all.js'
options    =  {float: false};
this.grid  	=   GridStack.init(options);

show Error in mounted hook: "TypeError: _gridstackAll.default.init is not a function"

How can I use it correctly？ thanks

**Comment by fuyouping** [comment]
this way

npm install gridstack
import GridStack from 'gridstack/dist/gridstack.all.js'

The result is the same as above

**Comment by tmpduarte** [comment]
Same here, there is also a [Stackoverflow issue](https://stackoverflow.com/questions/60602413/how-to-correctly-import-gridstack-using-npm-webpack) posing what seems to be the exact same problem

**Comment by tmpduarte** [comment]
a hack I found is to import the following:
```
import 'gridstack/dist/jquery';
import 'gridstack/dist/gridstack';
import 'gridstack/dist/jquery-ui';
import 'gridstack/dist/gridstack.jQueryUI';
```

**Comment by adumesny** [comment]
Hey guys, yeah there is likely something I'm doing wrong when generating `gridstack.all.js` (grunt job uglify to combine all files into 1
```js
'dist/gridstack.all.js': ['src/gridstack-poly.js', 'src/jquery.js', 'src/gridstack.js', 'src/jquery-ui.js', 'src/gridstack.jQueryUI.js']
```

if someone has any build export / webpack expertise I would appreciate it the help as I've been foxusing on 2.x (typescript re-write and removing jquery from main code) and could use packing help there as well. Let's fix 1.1.1 first!

**Comment by fuyouping** [comment]
> a hack I found is to import the following:
> 
> ```
> import 'gridstack/dist/jquery';
> import 'gridstack/dist/gridstack';
> import 'gridstack/dist/jquery-ui';
> import 'gridstack/dist/gridstack.jQueryUI';
> ```
thanks。i,m try try

**Comment by adumesny** [comment]
in my angular app it does work doing
`import 'gridstack/dist/gridstack.all.js';` 
BUT then none of the symbols from .d.ts are found, and doing that **instead**:
`import { GridStack, GridStackElement, GridStackNode, GridstackWidget } from 'gridstack';`
fixes the defines but then other code is missing $ jquery stuff (as it doesn't import jq/jq-ui) and doing both complains that `init()` doesn't exist on null.
Yeah so not ideal...

**Comment by fuyouping** [comment]
<template>
  	<div class="app-container gridLayoutContainer" id="gridLayoutContainer">
  		<div class="container-fluid">
	        <h1>Two grids demo</h1>
	        <div class="row">
	            <div>
	                <div class="sidebar">
	                    <div class="grid-stack-item">
	                        <div class="grid-stack-item-content"><button onClick="window.grid.removeWidget(this.parentNode.parentNode)">X</button><br>Drag me</div>
	                    </div>
	                </div>
	            </div>
	        </div>
	        <div class="row">
	            <div class="col-md-6">
	                <div class="grid-stack grid-stack-6"></div>
	            </div>
	        </div>
	        <a @click="saveGrid()" class="btn btn-primary" href="#">Save</a>
	    </div>
  	</div>
</template>
<style type="text/css">
	.sidebar{
		display: flex;
		flex-direction: row;
		height: 100px;
		line-height: 100px;
	}
	.sidebar .grid-stack-item{
		width: 100px;
		height: 50px;
		line-height: 50px;
		margin-left: 20px;
		z-index: 2;
	}
	.row{
		margin-top: 20px;
	}
</style>
<script>
import $ from 'gridstack/dist/jquery';
import 'gridstack/dist/gridstack';
import 'gridstack/dist/jquery-ui';
import 'gridstack/dist/gridstack.jQueryUI';
export default {
	name: '',
	data() {
		return {
			grid: 0,
		}
	},
	mounted() {
		this.init();
	},
 	methods: {
 		init(){
 			console.log(window.abc);
 			let that 	=	this;
 			var options = {
		        column: 6,
		        maxRow: 4,
		        minRow: 4,
		        float: true,
		        disableResize : true,
		        acceptWidgets: function(i, el) {return true; }
		    };
		    window.grid =	that.grid  	=   GridStack.init(options);
		    var items 	=   [
		        {x: 0, y: 0, width: 2, height: 2},
		        {x: 3, y: 1, width: 1, height: 2},
		        {x: 4, y: 1, width: 1, height: 1},
		        {x: 3, y: 3, width: 2, height: 1, id: 'special', text: 'has minWidth=2'},
		        {x: 2, y: 5, width: 1, height: 1}
		    ];

		    items.forEach(function(node){
		        that.grid.batchUpdate();
		        that.grid.addWidget('<div><div class="grid-stack-item-content"><button onClick="window.grid.removeWidget(this.parentNode.parentNode)">X</button><br>' + (node.text? node.text : '') + '</div></div>', node);
		        that.grid.commit();
		    });

		    $('.sidebar .grid-stack-item').draggable({
		        revert: 'invalid',
		        handle: '.grid-stack-item-content',
		        scroll: false,
		        appendTo: 'body',
		    });

		    that.grid.on('removed', function(event, items) {
		        console.log(event);
		        console.log(items);
		    });
 		},
 		saveGrid() {
 			let that 	=	this;
	        var serializedData = [];
	        that.grid.engine.nodes.forEach(function(node) {
	            serializedData.push({
	                x: node.x,
	                y: node.y,
	                width: node.width,
	                height: node.height
	            });
	        });
	        console.log(serializedData);
	    }
	}
}
</script>
<style>
@import 'https://gridstackjs.com/demo/demo.css';
@import 'https://cdn.jsdelivr.net/npm/gridstack@1.1.1/dist/gridstack-extra.min.css';
</style>


thinks!  its work.

**Comment by adumesny** [comment]
so your code works ? can I close this now ?
 Tip: in your code you should put the `batchUpdate()` and `commit()` **outside** of the foreach loop! (defeats the purpose otherwise).

**Comment by richard-ob** [comment]
> in my angular app it does work doing
> `import 'gridstack/dist/gridstack.all.js';`
> BUT then none of the symbols from .d.ts are found, and doing that **instead**:
> `import { GridStack, GridStackElement, GridStackNode, GridstackWidget } from 'gridstack';`
> fixes the defines but then other code is missing $ jquery stuff (as it doesn't import jq/jq-ui) and doing both complains that
> `ERROR TypeError: Cannot read property 'init' of undefined`
> Yeah so not ideal...

I'm using GridStack within an Angular app, however the second option (`import { GridStack, GridStackElement, GridStackNode, GridstackWidget } from 'gridstack';`) doesn't work for me? I'm able to use GridStack via importing it as a script in Angular CLI, however I want to avoid that, and import in the usual way, including allowing the usage of typings. Presently I'm using the `declare var GridStack`  hack.

I would raise an issue, but as you've mentioned this here, thought it made sense just to follow it up on this post. Thank you!

**Comment by adumesny** [comment]
@richard-ob thanks for adding here. Hopefully this will get resolved soon with 2.x

**Comment by jasonatball** [comment]
Tried importing this in a Vue.js project and ran into the same issue. Tried this:
```
import 'gridstack/dist/jquery'
import 'gridstack/dist/gridstack'
import 'gridstack/dist/jquery-ui'
import 'gridstack/dist/gridstack.jQueryUI'
export default {
  created() {
    var grid = GridStack.init()
    console.log('grid:' + grid)
  }
}
```
And the grid was 'undefined' but apparently it loaded because I got no error on GridStack.init().

**Comment by adumesny** [comment]
@jasonatball any error log messages ? make sure you have a div with `'.grid-stack'` class (the latest code will warn you if missing).

**Comment by jasonatball** [comment]
This was my misunderstanding of the Vue.js lifecycle. I needed to put the GridStack.init() in the mounted() lifecycle hook because until then there is no grid-stack div to even initialize against. Sorry for the noise.

**Comment by Marius8881** [comment]
Same issue with React...

**Comment by timothymarois** [comment]
Why is this closed? This issue is still a problem. Vue JS here.

**Comment by adumesny** [comment]
I suggest you use 2.0.0 (rc2 right now) as it's ES6 friendly (re-wrote to typescript) and look at 
https://github.com/gridstack/gridstack.js/tree/typescript#include
https://github.com/gridstack/gridstack.js/tree/typescript#migrating-to-v2

I have no plans to release 1.x again - on-ward and forward. Will close this as soon as v2 is officially out.

**Comment by adumesny** [comment]
this now works in just released 1.2.1 (see release notes)
```js
import 'jquery';
import 'jquery-ui';

import 'gridstack/dist/gridstack-poly'; // optional IE support
import GridStack from 'gridstack';
import 'gridstack/dist/gridstack.jQueryUI';
import 'gridstack/dist/gridstack.css';
```
but recommend you just use 2.x (native typescript re-write) which only needs this:
```js
import { GridStack, $ } from 'gridstack'; // $ optional if you depend on it, to be removed in 3.x
import 'gridstack/dist/gridstack.css';
```
might want to also see https://github.com/gridstack/gridstack.js#jquery-application
